File: po/po_limiter/page/po_limiter/po_limiter.py
# Copyright (c) 2026, Lassod
# License: MIT

import frappe
from frappe import _

def get_context(context):
	"""Get context for the PO Limiter page"""
	# Ensure only MD can access this page
	if not has_md_access():
		frappe.throw(_("You don't have permission to access this page."), frappe.PermissionError)

	context.no_cache = 1
	context.title = _("PO Limiter Tool")

	# Get all companies
	context.companies = frappe.get_all("Company", filters={"enabled": 1}, fields=["name", "abbr"])

	# Get users with PO access
	context.users = get_purchase_users()

	# Get all user limits
	context.user_limits = get_all_user_limits()

	# Get pending requests
	context.pending_requests = get_pending_limit_requests()

	return context


@frappe.whitelist()
def get_purchase_users():
	"""Get all users who have access to create Purchase Orders"""
	# Get users with Purchase Order create permission
	users = frappe.get_all("User", filters={
		"enabled": 1,
		"user_type": "System User"
	}, fields=["name", "full_name", "email"])

	purchase_users = []
	for user in users:
		if has_po_create_permission(user.name):
			purchase_users.append(user)

	return purchase_users


@frappe.whitelist()
def has_po_create_permission(user=None):
	"""Check if user has permission to create Purchase Orders"""
	if not user:
		user = frappe.session.user

	# Check if user has "Purchase Order Creator" or "Purchase Order Manager" role
	# or has create permission on Purchase Order
	has_role = frappe.db.exists("Has Role", {
		"parent": user,
		"role": ["in", ["Purchase Order Creator", "Purchase Order Manager", "System Manager", "Managing Director"]]
	})

	return has_role


@frappe.whitelist()
def get_all_user_limits():
	"""Get all user PO limits"""
	limits = frappe.get_all("User PO Limit",
		fields=["name", "user", "company", "status", "per_po_limit", "per_month_limit",
				"monthly_usage", "last_reset_date", "last_updated_by", "last_updated_date"],
		order_by="user, company"
	)
	return limits


@frappe.whitelist()
def get_pending_limit_requests():
	"""Get all pending PO limit increase requests"""
	requests = frappe.get_all("PO Limit Increase Request",
		filters={"status": "Pending Approval"},
		fields=["name", "user", "company", "requested_per_po_limit", "requested_per_month_limit",
				"reason", "creation"]
	)
	return requests


@frappe.whitelist()
def update_user_limit(user, company, per_po_limit, per_month_limit, status):
	"""Update or create user PO limit"""
	if not has_md_access():
		frappe.throw(_("You don't have permission to perform this action."), frappe.PermissionError)

	# Check if limit exists
	exists = frappe.db.exists("User PO Limit", {"user": user, "company": company})

	if exists:
		# Update existing limit
		frappe.db.set_value("User PO Limit", exists, {
			"per_po_limit": per_po_limit,
			"per_month_limit": per_month_limit,
			"status": status,
			"last_updated_by": frappe.session.user,
			"last_updated_date": frappe.utils.now()
		})
	else:
		# Create new limit
		frappe.get_doc({
			"doctype": "User PO Limit",
			"user": user,
			"company": company,
			"per_po_limit": per_po_limit,
			"per_month_limit": per_month_limit,
			"status": status,
			"monthly_usage": 0,
			"last_reset_date": frappe.utils.today()
		}).insert()

	frappe.msgprint(_("PO Limit updated for {0}").format(user))
	return {"success": True}


@frappe.whitelist()
def get_user_limit_details(user, company):
	"""Get user limit details for editing"""
	if not has_md_access():
		frappe.throw(_("You don't have permission to access this information."), frappe.PermissionError)

	limit = frappe.db.get_value("User PO Limit",
		{"user": user, "company": company},
		["name", "per_po_limit", "per_month_limit", "status", "monthly_usage",
		 "last_reset_date", "last_updated_by", "last_updated_date"],
		as_dict=1
	)

	return limit


def has_md_access():
	"""Check if current user has MD access"""
	# Check for Managing Director role
	has_md_role = frappe.db.exists("Has Role", {
		"parent": frappe.session.user,
		"role": "Managing Director"
	})

	# Also allow System Manager
	has_sm_role = frappe.db.exists("Has Role", {
		"parent": frappe.session.user,
		"role": "System Manager"
	})

	return has_md_role or has_sm_role

File: po/po_limiter/page/po_limiter/po_limiter.js
// Copyright (c) 2026, Lassod
// License: MIT

frappe.pages['po-limiter'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'PO Limiter Tool',
		single_column: true
	});

	// Load the HTML content
	$(page.main).find('.page-content').html(
		frappe.render_template('po_limiter', {
			companies: [],
			users: [],
			user_limits: [],
			pending_requests: []
		})
	);

	// Load initial data
	load_data(page);

	// Setup event handlers
	setup_event_handlers(page);
};

function load_data(page) {
	// Show loading
	page.main.find('.po-limiter-container').html('<div class="text-center">Loading...</div>');

	// Get data from server
	frappe.call({
		method: 'po.po_limiter.page.po_limiter.po_limiter.get_context',
		args: {},
		callback: function(r) {
			if (r.message) {
				// Re-render with data
				var context = r.message;
				$(page.main).find('.po-limiter-container').replaceWith(
					frappe.render_template('po_limiter', context)
				);

				// Re-setup event handlers after re-render
				setup_event_handlers(page);
			}
		}
	});
}

function setup_event_handlers(page) {
	var container = page.main.find('.po-limiter-container');

	// User selection change
	container.find('#user-select').on('change', function() {
		var user = $(this).val();
		if (user) {
			// Show company selection
			container.find('.company-selection').show();

			// Store selected user
			container.data('selected-user', user);
		} else {
			container.find('.company-selection').hide();
			container.find('#limit-editor').hide();
		}
	});

	// Company selection change
	container.find('#company-select').on('change', function() {
		var user = container.data('selected-user');
		var company = $(this).val();

		if (user && company) {
			// Load user's current limit
			load_user_limit(user, company);
		} else {
			container.find('#limit-editor').hide();
		}
	});

	// Save limit button
	container.find('#save-limit-btn').on('click', function() {
		var user = container.data('selected-user');
		var company = container.find('#company-select').val();
		var status = container.find('#limit-status').val();
		var per_po_limit = container.find('#per-po-limit').val();
		var per_month_limit = container.find('#per-month-limit').val();

		if (!user || !company) {
			frappe.msgprint('Please select a user and company');
			return;
		}

		// Save limit
		frappe.call({
			method: 'po.po_limiter.page.po_limiter.po_limiter.update_user_limit',
			args: {
				user: user,
				company: company,
				per_po_limit: per_po_limit,
				per_month_limit: per_month_limit,
				status: status
			},
			callback: function(r) {
				if (r.message && r.message.success) {
					// Reload data
					load_data(page);
				}
			}
		});
	});

	// Reset button
	container.find('#reset-limit-btn').on('click', function() {
		container.find('#limit-status').val('Active');
		container.find('#per-po-limit').val('0');
		container.find('#per-month-limit').val('0');
		container.find('#current-usage').hide();
	});

	// Edit limit from all limits table
	container.find('.btn-edit-limit').on('click', function() {
		var user = $(this).data('user');
		var company = $(this).data('company');

		// Switch to by-user tab
		container.find('a[href="#by-user"]').tab('show');

		// Set user and company
		container.find('#user-select').val(user);
		container.data('selected-user', user);
		container.find('.company-selection').show();
		container.find('#company-select').val(company);

		// Load limit details
		load_user_limit(user, company);
	});

	// Approve request button
	container.find('.btn-approve').on('click', function() {
		var request_name = $(this).data('request');

		frappe.confirm('Approve this limit increase request?', function() {
			frappe.call({
				method: 'po.po_limiter.doctype.po_limit_increase_request.po_limit_increase_request.approve_request',
				args: {
					request_name: request_name
				},
				callback: function(r) {
					if (r.message) {
						load_data(page);
					}
				}
			});
		});
	});

	// Reject request button
	container.find('.btn-reject').on('click', function() {
		var request_name = $(this).data('request');

		frappe.prompt([
			{
				fieldname: 'reason',
				fieldtype: 'Text',
				label: 'Rejection Reason',
				reqd: 1
			}
		], function(values) {
			frappe.call({
				method: 'po.po_limiter.doctype.po_limit_increase_request.po_limit_increase_request.reject_request',
				args: {
					request_name: request_name,
					rejection_reason: values.reason
				},
				callback: function(r) {
					if (r.message) {
						load_data(page);
					}
				}
			});
		}, 'Reject Request', 'Reject');
	});
}

function load_user_limit(user, company) {
	var container = $('.po-limiter-container');

	frappe.call({
		method: 'po.po_limiter.page.po_limiter.po_limiter.get_user_limit_details',
		args: {
			user: user,
			company: company
		},
		callback: function(r) {
			if (r.message) {
				var limit = r.message;

				// Populate form
				container.find('#limit-status').val(limit.status || 'Active');
				container.find('#per-po-limit').val(limit.per_po_limit || 0);
				container.find('#per-month-limit').val(limit.per_month_limit || 0);

				// Show current usage
				if (limit.monthly_usage) {
					container.find('#monthly-usage').text(
						frappe.format(limit.monthly_usage, {fieldtype: 'Currency'})
					);
					container.find('#current-usage').show();
				}

				container.find('#limit-editor').show();
			} else {
				// No existing limit - show empty form
				container.find('#limit-status').val('Active');
				container.find('#per-po-limit').val(0);
				container.find('#per-month-limit').val(0);
				container.find('#current-usage').hide();
				container.find('#limit-editor').show();
			}
		}
	});
}
